src/System.Private.CoreLib/shared/System/SpanHelpers.Byte.cs

```diff
@@ -255,29 +268,124 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
                 index += 1;
             }
 
-            if (Vector.IsHardwareAccelerated && ((int)(byte*)index < length))
+            if (Avx2.IsSupported)
             {
-                nLength = (IntPtr)((length - (int)(byte*)index) & ~(Vector<byte>.Count - 1));
+                if ((int)(byte*)index < length)
+                {
+                    nLength = (IntPtr)((length - (int)(byte*)index) & ~(Vector256<byte>.Count - 1));
+                    if ((byte*)nLength > (byte*)index)
+                    {
+                        Vector256<byte> comparison = Vector256.Create(value);
+                        do
+                        {
+                            Vector256<byte> search = Unsafe.ReadUnaligned<Vector256<byte>>(ref Unsafe.AddByteOffset(ref searchSpace, index));
```

Comment:
Why `ReadUnaligned` rather than `LoadVector256`? Is it to avoid pinning?

Reply:
No pinning here... :)

Its over raw byte data which is generally the longest type of data, so probably best to not get in the way of the GC; if it wants to compact the heap, rather than create a giant fixed block.

Reply:
Hmmm, the initial assumption was that `HWIntrinsics` would likely be used with longer inputs and, in those cases, that pinning would be desirable (so alignment, cache-coherency, etc could be preserved).

If that isn't the case, we might want to revisit whether or not we also expose `Load` overloads that take `ref T` rather than just `T*`.

Reply:
That's probably still the case for directly using them; but everything is getting plumbed though the SpanHelper methods so they are generic methods for all data sizes. e.g. string, array, span all use SpanHelpers now rather than their own implementations.

Reply:
Fixing the data would be required to ensure that the alignment check remains valid.

The user can pin the data if they want the alignment to remain fixed (though they may non-know to do so); Kestrel's arrays are already pre-pinned so a second pin here would add no advantage, native memory pinning here would add no advantage, lengths < 32 pinning would add no advantage; but in all those cases it would be extra unnecessary work to fix the data (GC considerations aside).

Is there a significant performance benefit to using `LoadVector256` over `ReadUnaligned` here?

Otherwise the unalignment issue will only effect the method if GC moves the data; and then, well you just had a GC so that will probably have more impact than being misaligned afterwards; and its also likely a low frequency event vs the number of times these methods are called.

Reply:
> Is there a significant performance benefit to using LoadVector256 over ReadUnaligned here?

Probably not, at least with the current RyuJIT...

Reply:
> >Is there a significant performance benefit to using LoadVector256 over ReadUnaligned here?
>
> Probably not, at least with the current RyuJIT...

I don't think so either, but if this style of code is frequent enough, it may be reason to have a `LoadVector256<T>(ref T)` overload (and maybe `LoadVector256<T>(ref T, offset)` as well). This would significantly help readability for this type of code since we would go from:
```diff
- Vector256<byte> search = Unsafe.ReadUnaligned<Vector256<byte>>(ref Unsafe.AddByteOffset(ref searchSpace, index));
+ Vector256<byte> search = Avx.LoadVector256(ref searchSpace, index);
```

Reply:
Ahh, see what you are saying, more about the parameters than the alignment.

Yes that would be nicer

Reply:
What would the index argument of the API be? We do not have native int public type yet...

Reply:
> What would the index argument of the API be? We do not have native int public type yet...

Presumably just `IntPtr`, as the existing `AddByteOffset` method takes: `public static ref T AddByteOffset<T>(ref T source, IntPtr byteOffset)`.

This emits an IL signature using `native int` and works correctly with languages (like F#) that treat `IntPtr` as a native sized integer. This should also work with C#, unless they decide to create a new wrapper type (rather than use partial erasure) or unless they decide that the `nint` type should have a `modreq`/`modopt`.

Reply:
> This should also work with C#, unless they decide to create a new wrapper type

I do not think there was a decision made on this one yet.

Reply:
Right, what C# does is still pending further language discussion. However, we have already exposed a number of similar APIs that take `IntPtr` (mostly on `S.R.CS.Unsafe`), and the worst case scenario is that we may want to expose an additional convenience overload (likely just implemented in managed code) that casts from whatever C# uses to `IntPtr`.

It could also just be internal for the time being, to make our own code more readable.

Reply:
Added local `LoadVector256`, `LoadVector128`, `LoadVector` and `LoadUIntPtr` to see how it works; makes it more readable.